// @picovoice/leopard-node/dist/errors.js
//
// Copyright 2022-2023 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
// file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.pvStatusToException = exports.LeopardActivationRefusedError = exports.LeopardActivationThrottledError = exports.LeopardActivationLimitReachedError = exports.LeopardActivationError = exports.LeopardRuntimeError = exports.LeopardInvalidStateError = exports.LeopardKeyError = exports.LeopardStopIterationError = exports.LeopardInvalidArgumentError = exports.LeopardIOError = exports.LeopardOutOfMemoryError = exports.LeopardError = void 0;
const pv_status_t_1 = require("./pv_status_t");
class LeopardError extends Error {
    _message;
    _messageStack;
    constructor(message, messageStack = []) {
        super(LeopardError.errorToString(message, messageStack));
        this._message = message;
        this._messageStack = messageStack;
    }
    get message() {
        return this._message;
    }
    get messageStack() {
        return this._messageStack;
    }
    static errorToString(initial, messageStack) {
        let msg = initial;
        if (messageStack.length > 0) {
            msg += `: ${messageStack.reduce((acc, value, index) => acc + '\n  [' + index + '] ' + value, '')}`;
        }
        return msg;
    }
}
exports.LeopardError = LeopardError;
class LeopardOutOfMemoryError extends LeopardError {
}
exports.LeopardOutOfMemoryError = LeopardOutOfMemoryError;
class LeopardIOError extends LeopardError {
}
exports.LeopardIOError = LeopardIOError;
class LeopardInvalidArgumentError extends LeopardError {
}
exports.LeopardInvalidArgumentError = LeopardInvalidArgumentError;
class LeopardStopIterationError extends LeopardError {
}
exports.LeopardStopIterationError = LeopardStopIterationError;
class LeopardKeyError extends LeopardError {
}
exports.LeopardKeyError = LeopardKeyError;
class LeopardInvalidStateError extends LeopardError {
}
exports.LeopardInvalidStateError = LeopardInvalidStateError;
class LeopardRuntimeError extends LeopardError {
}
exports.LeopardRuntimeError = LeopardRuntimeError;
class LeopardActivationError extends LeopardError {
}
exports.LeopardActivationError = LeopardActivationError;
class LeopardActivationLimitReachedError extends LeopardError {
}
exports.LeopardActivationLimitReachedError = LeopardActivationLimitReachedError;
class LeopardActivationThrottledError extends LeopardError {
}
exports.LeopardActivationThrottledError = LeopardActivationThrottledError;
class LeopardActivationRefusedError extends LeopardError {
}
exports.LeopardActivationRefusedError = LeopardActivationRefusedError;
function pvStatusToException(pvStatus, errorMessage, messageStack = []) {
    switch (pvStatus) {
        case pv_status_t_1.default.OUT_OF_MEMORY:
            throw new LeopardOutOfMemoryError(errorMessage, messageStack);
        case pv_status_t_1.default.IO_ERROR:
            throw new LeopardIOError(errorMessage, messageStack);
        case pv_status_t_1.default.INVALID_ARGUMENT:
            throw new LeopardInvalidArgumentError(errorMessage, messageStack);
        case pv_status_t_1.default.STOP_ITERATION:
            throw new LeopardStopIterationError(errorMessage, messageStack);
        case pv_status_t_1.default.KEY_ERROR:
            throw new LeopardKeyError(errorMessage, messageStack);
        case pv_status_t_1.default.INVALID_STATE:
            throw new LeopardInvalidStateError(errorMessage, messageStack);
        case pv_status_t_1.default.RUNTIME_ERROR:
            throw new LeopardRuntimeError(errorMessage, messageStack);
        case pv_status_t_1.default.ACTIVATION_ERROR:
            throw new LeopardActivationError(errorMessage, messageStack);
        case pv_status_t_1.default.ACTIVATION_LIMIT_REACHED:
            throw new LeopardActivationLimitReachedError(errorMessage, messageStack);
        case pv_status_t_1.default.ACTIVATION_THROTTLED:
            throw new LeopardActivationThrottledError(errorMessage, messageStack);
        case pv_status_t_1.default.ACTIVATION_REFUSED:
            throw new LeopardActivationRefusedError(errorMessage, messageStack);
        default:
            // eslint-disable-next-line no-console
            console.warn(`Unmapped error code: ${pvStatus}`);
            throw new LeopardError(errorMessage);
    }
}
exports.pvStatusToException = pvStatusToException;
